### ISSUE TITLE
Feature request: SSO support

### ISSUE THREAD

**Issue by skjdghsdjgsdj** [babybuddy/babybuddy#907]
You probably saw this one coming. I'm not sure if it's the same as #517.

I've been amassing a set of self-hosted services for my network, among which is Baby Buddy. In doing so, that means accumulating logins. While a password manager helps keep track of things, it doesn't help with management: what if I want to add a new user and give them access to all the things? To that end, I set up [Authentik](https://goauthentik.io/) as an SSO. I've successfully integrated it with some services like [Immich](https://immich.app/) and integrating it with Baby Buddy would be nice.

Example use case: I want to invite a family member to access all my family-related stuff. If my self-hosted services include a photo gallery, a family calendar, a recipe manager, Baby Buddy, etc., how can I give them a single login for all the things and that keeps them logged in as they navigate around? SSO is obviously the textbook use case for that.

Letting the SSO handle authentication also gets us some features for "free" like TOTP codes, passkeys, password reset flows, en masse user deactivations, etc. None of those would need to be implemented in Baby Buddy proper if the SSO did them.

Some caveats/notes:

* I believe Authentik can solve as a generic authenticator for a reverse proxy, but that's a band-aid solution to actual user management. Users logging in via Authentik or any other compatible SSO should result in actual user entities in Baby Buddy instead of just merely granting access.
* API access must continue to function. Would that still work as an API key associated to a user or proper service accounts (or both)?
* Would this affect user roles? Would the SSO, Baby Buddy, or both define actual authorization in addition to authentication?
* Would we allow both SSO and local logins?
* An average user likely doesn't want SSO. A default Baby Buddy install should still use local authentication. An administrator can add SSO login, or replace the built-in authentication with SSO, if they want.

This is probably a roadmap thing instead of a simple feature to implement unless it's [easy with Django](https://pypi.org/project/django-sso/). I am also far from versed on all the details of how SSO actually gets implemented, but it's obviously much more complex than the built-in Django authentication/authorization.

As I obviously have a functional SSO set up, I'd be happy to set up a test instance of Baby Buddy with bleeding edge code or a separate branch to test prior to an actual release.

**Comment by cdubz** [comment]
I think you should be able to use the existing functionality from #517 for this. I'm a little unclear on "proxy provider" note in their documentation, but this seems to imply it's possible to do: https://docs.goauthentik.io/docs/users-sources/user/user_ref#additionalheaders

> I believe Authentik can solve as a generic authenticator for a reverse proxy, but that's a band-aid solution to actual user management. Users logging in via Authentik or any other compatible SSO should result in actual user entities in Baby Buddy instead of just merely granting access.

Yes. This is what #517 does, if that path works. See: https://docs.djangoproject.com/en/5.1/howto/auth-remote-user/

> API access must continue to function. Would that still work as an API key associated to a user or proper service accounts (or both)? In lieu of full-on service accounts within Baby Buddy, I think for now it's okay to tie API keys to users of any kind.

Yep -- this should still work!

> Would this affect user roles? Would the SSO, Baby Buddy, or both define actual authorization in addition to authentication? Traditionally, the SSO provides authentication but the app (Baby Buddy) does the authorization; that is, the SSO says "user X is allowed to access Baby Buddy" but Baby Buddy says "user X is allowed to do Y within Baby Buddy". There are ways of doing both but that's too complicated for now I'd suggest.

No affect on user roles. As you suggest, the SSO will handle authentication, and Baby Buddy will continue to handle authorization.

> Would we allow both SSO and local logins? I would hope yes, but I'm sure some users would prefer to lock down local logins once an SSO is set up.

Both are still supported. This could be a separate change/FR to add a flag to disable local login.

> An average user likely doesn't want SSO. A default Baby Buddy install should still use local authentication.

And it does! 😁

**Comment by skjdghsdjgsdj** [comment]
I'll try this with a proxy later. I just have to be super certain that Baby Buddy can't be accessed without going through the proxy or an end user can spoof the authorized user header and log in as anyone!

What about longer-term a more native solution? That is, making Baby Buddy aware of the SSO's existence complete with the proper redirects and whatnot instead of just relying on magic HTTP headers and a proxy? I'm already proxying Baby Buddy through [Caddy](https://caddyserver.com/) so a proxy through Authentik on top of that adds more opportunities for latency and failure points.

**Comment by cdubz** [comment]
> What about longer-term a more native solution? That is, making Baby Buddy aware of the SSO's existence complete with the proper redirects and whatnot instead of just relying on magic HTTP headers and a proxy? I'm already proxying Baby Buddy through [Caddy](https://caddyserver.com/) so a proxy through Authentik on top of that adds more opportunities for latency and failure points.

I'm iffy on this. I see this as a "pretty good" solution for a what I assume is a pretty rare use case. Interested to hear about your experience as you try this out though.

**Comment by skjdghsdjgsdj** [comment]
So I got it working with a proxy but there are drawbacks with this approach where a native integration is still better, in my opinion.

My former architecture: (internet) -> Caddy (reverse proxy) -> Baby Buddy (docker)
My new architecture: (internet) -> Caddy (reverse proxy) -> Authentik (proxy) -> Baby Buddy (docker)

I configured Authentik as a proxy and whitelisted the paths `/login/*`, `/static/*`, and `/api/*`. Protip: Authentik users will need to assign the provider manually to the default outpost for this to work.

Because Authentik's proxy sets an `X-Authentik-Username` header for authenticated requests, In Baby Buddy, I set these environment variables:

```
 - PROXY_HEADER=HTTP_X_AUTHENTIK_USERNAME
 - REVERSE_PROXY_AUTH=True
 ```
Net result: unauthenticated requests to Baby Buddy redirect to Authentik's login and upon successful completion, Baby Buddy loads and shows me logged in as the correct user.

Now for the drawbacks:

* As expected, users who aren't locally created in Baby Buddy can't log in via SSO because the `X-Authentik-Username` header must map to a local Baby Buddy account. That means users have to be manually provisioned into Baby Buddy and aren't created on the fly. The password of said manual accounts, presumably, can be random given it doesn't matter what it is so long as it's unguessable.
* Logout from Baby Buddy doesn't work, also unsurprisingly, given Baby Buddy is unaware of the SSO proxy. Logging out kicks the user back to the SSO which promptly redirect the user straight back to Baby Buddy, still logged in. This is somewhat fixable if we add an environment variable to allow a custom logout link URL and then people with SSOs can enter the URL of their SSO's logout.
* Local login is still possible, but not obvious how to a typical user: they have to explicitly navigate to `/login/`. The only way I can see around this is modifying the SSO's login form to include a link to Baby Buddy's login form directly to allow for a local option. It seems the more common UX pattern is to show a local login but below that a "Log in with SSO" button or something similar. That could still be implemented via an environment variable like the logout suggestion above.

So tl;dr Baby Buddy supports SSO logins but with some significant and, if you think about it, unsurprising drawbacks.

Obviously it goes back to your comment about the actual broader desirability of an SSO. I won't fool myself: I know most people won't use it, but it does fit my use case at least and I imagine at least a few other self-hosters who are building a collection of services on their network.

**Comment by skjdghsdjgsdj** [comment]
Caddy workaround (hacky) to get log out and password change working:

```
redir /logout/ https://authentik.domain/flows/-/default/invalidation/                                                         
redir /user/password/ https://authentik.domain/if/flow/default-password-change/
```

**Comment by skjdghsdjgsdj** [comment]
Another interesting thing: when proxied through Baby Buddy, the images for children are broken. They respond with far fewer bytes and a 404, even when I exempt `/media/*` from authentication in Authentik. I had to set up a separate proxy in Caddy directly to the Baby Buddy endpoint for `/media/*`.

**Comment by juantxorena** [comment]
I want to add my support to this, actually to oauth2 or oidc support. The aforementioned tandoor app uses django-allauth for this, which basically allows to use any SSO for login, either selfhosted or not (like google or github accounts). I don't know python, but tandoor has almost no code related to that, so maybe it's as easy as adding the dependency. The only thing would be in the frontend, I guess.

> An average user likely doesn't want SSO. A default Baby Buddy install should still use local authentication.

I disagree. Babybuddy is a selfhosted app, and in my experience, one of the first things any selfhoster does is to set up a SSO.

**Comment by juantxorena** [comment]
Because I really wanted this before my daughter is born, I took the liberty to "fix" it and open a pull request for SSO support: #973. There is still work to be done there. I would kindly ask for help in fixing the missing parts, and for help in testing.

**Comment by PedDavid** [comment]
@juantxorena , first of all, congrats on the daughter that most likely is already here 🥳 

Unfortunately I hadn't your foresight and I'm only looking into this after my son is already "out" so were you able to leverage Babybuddy with SSO from your branch?

I agree this should be much more native instead of relying on "proxy hacks".
More and more home-lab / self-hosted apps are now relying on SSO with the ease of deploy with Pocket ID and such (honestly just stating what I've seen, I actually use Keycloak) and SSO seems almost essential to sell the idea to non tech savvy family members and what not but the barrier to set it up with the proxy is much bigger (at least in my experience... and I'm using keycloak which many consider heavyweight / hard compared to newer alternatives...).
